Send DS's SDL to clients from the AuthSrv

Hoikas opened the pull request:
H-uru/Plasma#1450 changed the meaning of `/LocalData` to "use the local PRPs and Python files" but still rely on the server's SDL. This means that clients either need to use the new `/LocalSDL` flag to use their own SDL, which is known to be dangerous, or we need to make it easier to get the known correct SDL to clients connecting to DirtSand.

Therefore, this adds the ability for DirtSand to automatically send down the SDL it is using internally to clients. It uses the very old auth secure preloader functionality, so it is compatible with all known clients. Most SDL files are quite small, which results in fairly poor download performance. Further, AuthSrv downloads are not compressed. So, I elected to give DirtSand the ability to write out its own SDL file containing all known state descriptors and omitting irrelevant comments and whitespace to give an optimal experience for users.

The last two commits could be deferred to another pull request, but I suspected some of the excess copying and memory management might be remarked upon, so I included those ideas for completeness' sake.

Hoikas committed:
Add EncryptedStream and tests.

Hoikas committed:
Add support for reading encrypted SDL files.

The encrypted stream class in *intended* for writing out minified copies
of the SDL files for the client, but adding the ability to read
encrypted SDL is just such low-hanging fruit that we might as well add
support for it.

dgelessus commented on streams.cpp:
```diff
@@ -258,3 +300,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
+    for (size_t i = 0; i < 32; i++) {
+        second -= (((first >> 5) ^ (first << 4)) + first)
+                ^ (m_key[(key >> 11) & 3] + key);
+        key += 0x61C88647;
```
Could you flip the sign here so the magic number matches the client?

```suggestion
        key -= 0x9E3779B9;
```

dgelessus commented on streams.cpp:
```diff
@@ -258,3 +300,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
+    for (size_t i = 0; i < 32; i++) {
+        first += (((second >> 5) ^ (second << 4)) + second)
+               ^ (m_key[key & 3] + key);
+        key -= 0x61C88647;
```
```suggestion
        key += 0x9E3779B9;
```

Hoikas committed:
Change magic numbers to match the client exactly.

Co-authored-by: dgelessus <dgelessus@users.noreply.github.com>

zrax commented on SDL/DescriptorDb.cpp:
```diff
@@ -88,39 +90,154 @@ SDL::StateDescriptor* SDL::DescriptorDb::FindDescriptor(const ST::string& name,
+    }
+}
+
+void SDL::DescriptorDb::WriteDescriptors(DS::Stream* stream)
```
Yikes!  Why not just encrypt the source file, instead of adding all this additional complexity by writing out processed files?  More specifically, since you've added the capability to handle encrypted files, why not just use the encrypted files as a source rather than continuing to store decrypted SDL separately (hint: that's also what I did for moulars ;) )...

Hoikas replied:
I'm having trouble parsing the moulars source code atm, so I'm not really sure what's going on over there. I do agree, though, that this method is fairly complex. My goal is to ensure that the SDL that the AuthSrv sends to the client is functionally the same as what is read into the descriptor database. My primary goal with this and the client change is to enforce a contract about what the SDL is - the processing is just a side effect with some benefits and drawbacks (namely, the added complexity). My thinking was that if we don't re-encode the SDL that's in the database but rather simply read it off the disk each time, then we would have a problem with that contract. DS currently initializes the descriptor database at start up and doesn't bother listening for any changes to the SDL directory. If we just send the contents of the SDL directory to client, any changes to the SDL directory would get sent down to the client but be unknown to DS. That really defeats the purpose of this PR, IMO.

Big picture: My eventual goal is to, in future changes, allow DS to listen for changes to its SDL directory. Ideally, this will allow changes to state descriptor versions themselves so that Ages undergoing active development don't require a vault wipe or excessive SDL version increments for minor development churn (this has been a problem for some Age/content creators in the past). To effect this change, DS needs to know the same state descriptors that the client knows so that the two can communicate in the same blob format. Right now, DS isn't able to cope with two version 1s of the same state descriptor, so I thought that round-tripping the SDL would be for the best. There is already precedent for this - DS round-trips everything else, even Plasma messages that could just be broadcast without being checked.

> why not just use the encrypted files as a source

I did update the code to allow the SDL files to be encrypted, but it doesn't seem to me like that's the best default way to store things. Ideally, we just want the SDL to be plaintext on the server for ease of administration.

zrax commented on AuthServ/AuthServer_Private.h:
```diff
@@ -105,23 +106,20 @@ struct AuthServer_Private : public AuthClient_Private
-            delete item->second;
-            m_downloads.erase(item);
-        }
+        m_downloads.clear();
```
This should happen automatically in the `std::map` destructor, so this destructor can be fully default now.

Hoikas replied:
I wasn't sure if we wanted to retain explicit clear on destruct or not since the previous constructor did explicitly remove each element from the map after deleting the stream. Will fix.

Hoikas committed:
Use correct names for encryption algos.

Hoikas committed:
Send SDLs from the server SDL directory.

The client now prefers to download SDL files from the server, even when
`/LocalData` is supplied. Therefore, it is important that DS is always
able to serve SDL files without intervention. So, when SDLs are
requested from the AuthServ, send down an encrypted copy of the SDLs
that DS loaded.

Hoikas commented:
I've removed the round-trip minifying step from this PR. Encrypted SDL files are not required for ease of use - DS can encrypt files before sending them down.

zrax commented on SDL/DescriptorDb.h:
```diff
@@ -100,17 +100,21 @@ namespace SDL
         static StateDescriptor* FindDescriptor(const ST::string& name, int version);
         static StateDescriptor* FindLatestDescriptor(const ST::string& name);
         static bool ForLatestDescriptors(descfunc_t functor);
+        static bool ForDescriptorFiles(const char* sdlpath, const filefunc_t& functor);
```
```suggestion
        static bool ForDescriptorFiles(const char* sdlpath, filefunc_t functor);
```

zrax commented on AuthServ/AuthServer.cpp:
```diff
@@ -566,7 +570,33 @@ void cb_fileList(AuthServer_Private& client)
+            return true;
+        };
+        try {
+            SDL::DescriptorDb::ForDescriptorFiles(DS::Settings::SdlPath(), populateSdl);
```
```suggestion
            SDL::DescriptorDb::ForDescriptorFiles(DS::Settings::SdlPath(), std::move(populateSdl));
```

zrax commented on SDL/DescriptorDb.cpp:
```diff
@@ -124,3 +117,27 @@ bool SDL::DescriptorDb::ForLatestDescriptors(descfunc_t functor)
     return true;
 }
 
+bool SDL::DescriptorDb::ForDescriptorFiles(const char* sdlpath, const filefunc_t& functor)
```
```suggestion
bool SDL::DescriptorDb::ForDescriptorFiles(const char* sdlpath, filefunc_t functor)
```

zrax commented on AuthServ/AuthServer.cpp:
```diff
@@ -604,12 +635,18 @@ void cb_downloadStart(AuthServer_Private& client)
+        filename = DS::Settings::AuthRoot() + filename;
+    }
+
+    std::unique_ptr<DS::Stream> stream = std::make_unique<DS::FileStream>();
```
```suggestion
    auto stream = std::make_unique<DS::FileStream>();
```
A `unique_ptr<U>` can be converted to a `unique_ptr<T>` if `U` derives from `T`, so there's no need for the extra casting complexity...

Hoikas replied:
While this is true, later in the function, a `DS::BufferStream` can be moved to this `std::unique_ptr`. I'll do some work to remove the fiddly casting to make that more clear.

zrax commented on AuthServ/AuthServer.cpp:
```diff
@@ -604,12 +635,18 @@ void cb_downloadStart(AuthServer_Private& client)
+    std::unique_ptr<DS::Stream> stream = std::make_unique<DS::FileStream>();
     try {
-        stream->open(filename.c_str(), "rb");
+        static_cast<DS::FileStream*>(stream.get())->open(filename.c_str(), "rb");
```
```suggestion
        stream->open(filename.c_str(), "rb");
```

zrax commented on AuthServ/AuthServer.cpp:
```diff
@@ -618,10 +655,27 @@ void cb_downloadStart(AuthServer_Private& client)
 
+    // All auth downloads must be encrypted.
+    if (!DS::EncryptedStream::CheckEncryption(stream.get()).has_value()) {
+        std::unique_ptr<DS::Stream> bufStream = std::make_unique<DS::BufferStream>();
```
```suggestion
        auto bufStream = std::make_unique<DS::BufferStream>();
```

zrax commented on SDL/SdlParser.cpp:
```diff
@@ -27,25 +29,34 @@ static const char* s_toknames[] = {
+        m_fileStream->open(filename, "r");
+    } catch (DS::FileIOException& ex) {
+        ST::printf(stderr, "[SDL] Error opening file {} for reading: {}\n", filename, ex.what());
         return false;
```
Leak (the previous version used `fopen(...) == NULL`, which meant no leaked handle was present)

zrax commented on streams.h:
```diff
@@ -272,6 +275,59 @@ namespace DS
+        EncryptedStream(Stream* base, Mode mode, std::optional<Type> type = std::nullopt, const uint32_t* keys = nullptr);
+        ~EncryptedStream() override;
+
+        EncryptedStream(const EncryptedStream&) = delete;
+        EncryptedStream(EncryptedStream&&) = delete;
```
Probably also worth explicitly deleting the assignment operators

zrax commented on streams.cpp:
```diff
@@ -258,3 +279,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
+    switch (mode) {
+    case Mode::e_read:
+        base->readBytes(header, sizeof(header));
+        if (memcmp(header, "whatdoyousee", 12) == 0)
+            m_type = DS::EncryptedStream::Type::e_tea;
+        else if (memcmp(header, "BriceIsSmart", 12) == 0)
+            m_type = DS::EncryptedStream::Type::e_tea;
```
```suggestion
        if (memcmp(header, "whatdoyousee", 12) == 0 || memcmp(header, "BriceIsSmart", 12) == 0)
            m_type = DS::EncryptedStream::Type::e_tea;
```

zrax commented on streams.cpp:
```diff
@@ -258,3 +279,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
+                break;
+        }
+        m_base->write<uint32_t>(m_size);
+    }
```
I would probably prefer this to be in a separate function that is then called by the destructor...  Otherwise, the destructor is the only way to finalize the stream header.

zrax commented on streams.cpp:
```diff
@@ -258,3 +279,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
+    stream->readBytes(header, sizeof(header));
+    stream->seek(pos, SEEK_SET);
+
+    if (memcmp(header, "whatdoyousee", sizeof(header)) == 0)
+        return DS::EncryptedStream::Type::e_tea;
+    if (memcmp(header, "BriceIsSmart", sizeof(header)) == 0)
+        return DS::EncryptedStream::Type::e_tea;
```
```suggestion
    if (memcmp(header, "whatdoyousee", sizeof(header)) == 0 || memcmp(header, "BriceIsSmart", sizeof(header)) == 0)
        return DS::EncryptedStream::Type::e_tea;
```

Hoikas committed:
Address issues from CR.

Co-authored-by: Michael Hansen <zrax0111@gmail.com>
